src/metabase/query_analysis/native_query_analyzer.clj

```diff
@@ -285,12 +302,36 @@
     {:tables (strip-model-refs table-refs)
      :fields (strip-model-refs field-refs)}))
 
+(defn- tables-for-sql
+  "Returns a set of table identifiers that (may) be referenced in the given card's query.
+  Errs on the side of optimism: i.e., it may return tables that are *not* in the query, and is unlikely to fail
+  to return tables that are in the query."
+  [driver query & {:keys [mode] :or {mode :ast-walker-1}}]
+  (let [db-id        (:database query)
+        macaw-opts   (nqa.impl/macaw-options driver)
+        table-opts   (assoc macaw-opts :mode mode)
+        sql-string   (:query (nqa.sub/replace-tags query))
+        parsed-query (macaw/query->tables sql-string table-opts)]
+    (table-refs-for-query parsed-query db-id)))
+
 (defn references-for-native
   "Returns a `{:explicit #{...} :implicit #{...}}` map with field IDs that (may) be referenced in the given card's
   query. Currently only support SQL-based dialects."
   [query]
   (let [driver (driver.u/database->driver (:database query))]
     ;; TODO this approach is not extensible, we need to move to multimethods.
     ;; See https://github.com/metabase/metabase/issues/43516 for long term solution.
+    ;; For now we are not restricting this based on the multimethod, as its useful to gather info on which drivers
+    ;; have errors or inaccuracies in practice.
     (when (isa? driver/hierarchy driver :sql)
       (references-for-sql driver query))))
+
+(defn tables-for-native
+  "Returns a set of table identifiers that (may) be referenced in the given card's query.
+  Takes an options :mode option, which determines the complexity of queries it can handle, and what types of false
+  positives it may return."
+  [query & {:as opts}]
+  (let [driver (driver.u/database->driver (:database query))]
+    (if (supported? driver)
+      (tables-for-sql driver query opts)
```

Comment:
Something to consider in future:

Would we ever want to support different code paths for analysis on a per-driver basis? e.g. a third party Mongo query analyzer, or an alternative to Macaw?

I'm not sure the best way we'd define that. Perhaps just using a regular multimethod instead of driver support?

Reply:
Decided to go with a regular multi-method instead, so that the analyzer is totally pluggable.